<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>How to read paired-end reads from a BAM file • BiocHowTo</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><script src="how-to-read-paired-end-reads-from-bam-file_files/libs/quarto-html/popper.min.js"></script><script src="how-to-read-paired-end-reads-from-bam-file_files/libs/quarto-html/tippy.umd.min.js"></script><link href="how-to-read-paired-end-reads-from-bam-file_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="how-to-read-paired-end-reads-from-bam-file_files/libs/quarto-html/light-border.css" rel="stylesheet">
<!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="How to read paired-end reads from a BAM file">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">BiocHowTo</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item"><a class="nav-link" href="../articles/index.html">Articles</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-quarto">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>How to read paired-end reads from a BAM file</h1>

      <p class="author">Bioconductor Core Team</p>
      <p class="date">2025-01-25</p>


      <div class="d-none name"><code></code></div>
    </div>






    <p>This HowTo has been adapted from the list of HowTos provided in the <a href="https://bioconductor.org/packages/release/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesHOWTOs.pdf" class="external-link">vignette</a> for the <em><a href="https://bioconductor.org/packages/3.21/GenomicRanges" class="external-link">GenomicRanges</a></em> Bioconductor package.</p>
<section class="section level2"><h2 id="bioconductor-packages-used-in-this-document">Bioconductor packages used in this document<a class="anchor" aria-label="anchor" href="#bioconductor-packages-used-in-this-document"></a>
</h2>
<ul>
<li><em><a href="https://bioconductor.org/packages/3.21/pasillaBamSubset" class="external-link">pasillaBamSubset</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.21/GenomicAlignments" class="external-link">GenomicAlignments</a></em></li>
</ul></section><section class="section level2"><h2 id="how-to-read-paired-end-reads-from-a-bam-file">How to read paired-end reads from a BAM file<a class="anchor" aria-label="anchor" href="#how-to-read-paired-end-reads-from-a-bam-file"></a>
</h2>
<p>For illustration, we use data from the <em><a href="https://bioconductor.org/packages/3.21/pasillaBamSubset" class="external-link">pasillaBamSubset</a></em> data package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pasillaBamSubset</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co"># Path to a bam file with paired-end reads</span></span>
<span><span class="op">(</span><span class="va">un3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/pasillaBamSubset/man/pasillaBamSubset-package.html" class="external-link">untreated3_chr4</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/Users/runner/work/_temp/Library/pasillaBamSubset/extdata/untreated3_chr4.bam"</span></span></code></pre></div>
</div>
<p>Paired-end reads can be loaded with the <code><a href="https://rdrr.io/pkg/GenomicAlignments/man/readGAlignments.html" class="external-link">GenomicAlignments::readGAlignmentPairs</a></code> or <code><a href="https://rdrr.io/pkg/GenomicAlignments/man/readGAlignments.html" class="external-link">GenomicAlignments::readGAlignmentsList</a></code> functions from the <em><a href="https://bioconductor.org/packages/3.21/GenomicAlignments" class="external-link">GenomicAlignments</a></em> package. These functions use the same mate pairing algorithm but output different objects.</p>
<p>Let’s start with <code><a href="https://rdrr.io/pkg/GenomicAlignments/man/readGAlignments.html" class="external-link">GenomicAlignments::readGAlignmentPairs</a></code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicAlignments" class="external-link">GenomicAlignments</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">gapairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicAlignments/man/readGAlignments.html" class="external-link">readGAlignmentPairs</a></span><span class="op">(</span><span class="va">un3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">gapairs</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "GAlignmentPairs"</span></span>
<span><span class="co">#&gt; attr(,"package")</span></span>
<span><span class="co">#&gt; [1] "GenomicAlignments"</span></span>
<span><span class="va">gapairs</span></span>
<span><span class="co">#&gt; GAlignmentPairs object with 75409 pairs, strandMode=1, and 0 metadata columns:</span></span>
<span><span class="co">#&gt;           seqnames strand   :          ranges  --          ranges</span></span>
<span><span class="co">#&gt;              &lt;Rle&gt;  &lt;Rle&gt;   :       &lt;IRanges&gt;  --       &lt;IRanges&gt;</span></span>
<span><span class="co">#&gt;       [1]     chr4      +   :         169-205  --         326-362</span></span>
<span><span class="co">#&gt;       [2]     chr4      +   :         943-979  --       1086-1122</span></span>
<span><span class="co">#&gt;       [3]     chr4      +   :         944-980  --       1119-1155</span></span>
<span><span class="co">#&gt;       [4]     chr4      +   :         946-982  --        986-1022</span></span>
<span><span class="co">#&gt;       [5]     chr4      +   :        966-1002  --       1108-1144</span></span>
<span><span class="co">#&gt;       ...      ...    ... ...             ... ...             ...</span></span>
<span><span class="co">#&gt;   [75405]     chr4      -   : 1348209-1348245  -- 1348063-1348099</span></span>
<span><span class="co">#&gt;   [75406]     chr4      -   : 1348211-1348247  -- 1348077-1348113</span></span>
<span><span class="co">#&gt;   [75407]     chr4      +   : 1348217-1348253  -- 1348215-1348251</span></span>
<span><span class="co">#&gt;   [75408]     chr4      +   : 1349196-1349232  -- 1349326-1349362</span></span>
<span><span class="co">#&gt;   [75409]     chr4      +   : 1349708-1349744  -- 1349838-1349874</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 8 sequences from an unspecified genome</span></span></code></pre></div>
</div>
<p>The <code>GAlignmentPairs</code> class holds only pairs; reads with no mate or with ambiguous pairing are discarded. Each list element holds exactly 2 records (a mated pair). Records can be accessed as the first and last segments in a template or as left and right alignments. See <code><a href="https://rdrr.io/pkg/GenomicAlignments/man/GAlignmentPairs-class.html" class="external-link">?GAlignmentPairs</a></code> in the <em><a href="https://bioconductor.org/packages/3.21/GenomicAlignments" class="external-link">GenomicAlignments</a></em> package for more information.</p>
<p>For <code>readGAlignmentsList</code>, mate pairing is performed when <code>asMates</code> is set to <code>TRUE</code> on the <code>BamFile</code> object, otherwise records are treated as single-end.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">galist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicAlignments/man/readGAlignments.html" class="external-link">readGAlignmentsList</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html" class="external-link">BamFile</a></span><span class="op">(</span><span class="va">un3</span>, asMates <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">galist</span></span>
<span><span class="co">#&gt; GAlignmentsList object of length 96636:</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; GAlignments object with 2 alignments and 0 metadata columns:</span></span>
<span><span class="co">#&gt;       seqnames strand       cigar    qwidth     start       end     width</span></span>
<span><span class="co">#&gt;          &lt;Rle&gt;  &lt;Rle&gt; &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt;   [1]     chr4      +         37M        37       169       205        37</span></span>
<span><span class="co">#&gt;   [2]     chr4      -         37M        37       326       362        37</span></span>
<span><span class="co">#&gt;           njunc</span></span>
<span><span class="co">#&gt;       &lt;integer&gt;</span></span>
<span><span class="co">#&gt;   [1]         0</span></span>
<span><span class="co">#&gt;   [2]         0</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 8 sequences from an unspecified genome</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; GAlignments object with 2 alignments and 0 metadata columns:</span></span>
<span><span class="co">#&gt;       seqnames strand       cigar    qwidth     start       end     width</span></span>
<span><span class="co">#&gt;          &lt;Rle&gt;  &lt;Rle&gt; &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt;   [1]     chr4      +         37M        37       946       982        37</span></span>
<span><span class="co">#&gt;   [2]     chr4      -         37M        37       986      1022        37</span></span>
<span><span class="co">#&gt;           njunc</span></span>
<span><span class="co">#&gt;       &lt;integer&gt;</span></span>
<span><span class="co">#&gt;   [1]         0</span></span>
<span><span class="co">#&gt;   [2]         0</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 8 sequences from an unspecified genome</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt; GAlignments object with 2 alignments and 0 metadata columns:</span></span>
<span><span class="co">#&gt;       seqnames strand       cigar    qwidth     start       end     width</span></span>
<span><span class="co">#&gt;          &lt;Rle&gt;  &lt;Rle&gt; &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt;   [1]     chr4      +         37M        37       943       979        37</span></span>
<span><span class="co">#&gt;   [2]     chr4      -         37M        37      1086      1122        37</span></span>
<span><span class="co">#&gt;           njunc</span></span>
<span><span class="co">#&gt;       &lt;integer&gt;</span></span>
<span><span class="co">#&gt;   [1]         0</span></span>
<span><span class="co">#&gt;   [2]         0</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 8 sequences from an unspecified genome</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ...</span></span>
<span><span class="co">#&gt; &lt;96633 more elements&gt;</span></span></code></pre></div>
</div>
<p><code>GAlignmentsList</code> is a more general ‘list-like’ structure that holds mate pairs as well as nonmates (i.e., singletons, records with unmapped mates etc). A <code>mates_status</code> metadata column (accessed with <code>mcols</code>) indicates which records were paired.</p>
<p>Non-mated reads are returned as groups by <code>QNAME</code> and contain any number of records. Here the non-mate groups range in size from 1 to 9.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">non_mates</span> <span class="op">&lt;-</span> <span class="va">galist</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">galist</span><span class="op">)</span><span class="op">$</span><span class="va">mate_status</span><span class="op">)</span> <span class="op">==</span> <span class="st">"unmated"</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/List-class.html" class="external-link">elementNROWS</a></span><span class="op">(</span><span class="va">non_mates</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     1     2     3     4     5     6     7     8     9 </span></span>
<span><span class="co">#&gt; 18191  2888    69    60     7     8     2     1     1</span></span></code></pre></div>
</div>
</section><section class="section level2"><h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<details><summary><b> Click to display session info </b>
</summary><div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R Under development (unstable) (2025-01-24 r87627)</span></span>
<span><span class="co">#&gt; Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">#&gt; Running under: macOS Sonoma 14.7.2</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] GenomicAlignments_1.43.0    Rsamtools_2.23.1           </span></span>
<span><span class="co">#&gt;  [3] Biostrings_2.75.3           XVector_0.47.2             </span></span>
<span><span class="co">#&gt;  [5] SummarizedExperiment_1.37.0 Biobase_2.67.0             </span></span>
<span><span class="co">#&gt;  [7] MatrixGenerics_1.19.1       matrixStats_1.5.0          </span></span>
<span><span class="co">#&gt;  [9] GenomicRanges_1.59.1        GenomeInfoDb_1.43.2        </span></span>
<span><span class="co">#&gt; [11] IRanges_2.41.2              S4Vectors_0.45.2           </span></span>
<span><span class="co">#&gt; [13] BiocGenerics_0.53.3         generics_0.1.3             </span></span>
<span><span class="co">#&gt; [15] pasillaBamSubset_0.45.0     BiocStyle_2.35.0           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] Matrix_1.7-1            jsonlite_1.8.9          compiler_4.5.0         </span></span>
<span><span class="co">#&gt;  [4] BiocManager_1.30.25     crayon_1.5.3            bitops_1.0-9           </span></span>
<span><span class="co">#&gt;  [7] parallel_4.5.0          BiocParallel_1.41.0     yaml_2.3.10            </span></span>
<span><span class="co">#&gt; [10] fastmap_1.2.0           lattice_0.22-6          R6_2.5.1               </span></span>
<span><span class="co">#&gt; [13] S4Arrays_1.7.1          knitr_1.49              DelayedArray_0.33.3    </span></span>
<span><span class="co">#&gt; [16] GenomeInfoDbData_1.2.13 rlang_1.1.4             xfun_0.50              </span></span>
<span><span class="co">#&gt; [19] SparseArray_1.7.2       cli_3.6.3               digest_0.6.37          </span></span>
<span><span class="co">#&gt; [22] grid_4.5.0              evaluate_1.0.3          codetools_0.2-20       </span></span>
<span><span class="co">#&gt; [25] abind_1.4-8             rmarkdown_2.29          httr_1.4.7             </span></span>
<span><span class="co">#&gt; [28] tools_4.5.0             htmltools_0.5.8.1       UCSC.utils_1.3.0</span></span></code></pre></div>
</div>
</details></section><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'light-border',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config);
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script></main>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Charlotte Soneson, Hervé Pagès, Dan Tenenbaum, Valerie Obenchain, Marc Carlson, Martin Morgan, James Hester, Vince Carey.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer>
</div>





  </body>
</html>
